File: lifebook/ios-native/HughManatee/ViewModels/ConversationViewModel.swift
import SwiftUI

@Observable
final class ConversationViewModel {
    enum ConvStatus {
        case loading, ready, live, ending, error
    }

    var status: ConvStatus = .loading
    var statusText = "Finding Hugh…"
    var firstTurn: String?
    var error: String?
    var turnCount = 0

    private let db = DatabaseService.shared
    private let worker = WorkerAPI.shared
    private let liveKit = LiveKitService()

    private var currentSessionId: String?
    private var turnOrdinal = 0
    private var turnsBuffer: [(speaker: String, text: String)] = []
    private var started = false
    private var ended = false
    private var turnTask: Task<Void, Never>?
    private var agentTask: Task<Void, Never>?
    private var statusTask: Task<Void, Never>?

    func startSession() {
        guard !started, !ended else { return }
        started = true

        Task {
            do {
                guard let profile = try await db.getProfile() else {
                    await setError("No profile found")
                    return
                }

                let lastAnchor = try await db.getLastAnchor()

                let cfg = try await worker.fetchAgentConfig(AgentConfigRequest(
                    firstName: profile.firstName,
                    birthYear: profile.birthYear,
                    hometown: profile.hometown,
                    voiceId: profile.voiceId,
                    lastAnchor: lastAnchor
                ))

                await MainActor.run {
                    firstTurn = cfg.first_turn
                    status = .ready
                    statusText = "Opening the mic…"
                }

                // Create session in DB
                let session = try await db.createSession(promptVersion: "agent-2026-05-06")
                currentSessionId = session.id

                // Connect LiveKit
                await liveKit.connect(token: cfg.conversation_token)

                // Collect turns
                turnTask = Task {
                    for await turn in liveKit.turns {
                        turnsBuffer.append((turn.speaker, turn.text))
                        guard let sid = currentSessionId else { continue }
                        turnOrdinal += 1
                        try? await db.appendTurn(
                            sessionId: sid,
                            speaker: turn.speaker,
                            text: turn.text,
                            ordinal: turnOrdinal
                        )
                        await MainActor.run { turnCount = turnOrdinal }
                    }
                }

                // Watch agent speaking state
                agentTask = Task {
                    while !Task.isCancelled {
                        await MainActor.run {
                            if status == .live {
                                statusText = liveKit.agentSpeaking ? "Hugh is speaking." : "Hugh is listening."
                            }
                        }
                        try? await Task.sleep(for: .milliseconds(200))
                    }
                }

                // Wait for connection
                statusTask = Task {
                    // Brief delay for LiveKit to connect
                    try? await Task.sleep(for: .seconds(1))
                    await MainActor.run {
                        if liveKit.status == .connected {
                            status = .live
                            statusText = "Hugh is listening."
                        }
                    }
                }

            } catch {
                await setError(error.localizedDescription)
            }
        }
    }

    func endSession() {
        guard !ended else { return }
        ended = true

        Task {
            await MainActor.run {
                status = .ending
                statusText = "Saving the memory…"
            }

            turnTask?.cancel()
            agentTask?.cancel()
            statusTask?.cancel()

            await liveKit.disconnect()

            guard let sid = currentSessionId else { return }

            var title: String? = nil
            var anchor: String? = nil

            let tail = turnsBuffer.suffix(10)
            if !tail.isEmpty {
                let payload = AnchorRequest(turns: tail.map {
                    TurnPayload(speaker: $0.speaker, text: $0.text)
                })
                if let anchorRes = try? await worker.fetchSessionAnchor(payload) {
                    title = anchorRes.title_suggestion
                    anchor = anchorRes.anchor_phrase
                }
            }

            try? await db.endSession(id: sid, title: title, anchorPhrase: anchor)
        }
    }

    func setMuted(_ muted: Bool) {
        liveKit.setMuted(muted)
    }

    @MainActor
    private func setError(_ msg: String) {
        status = .error
        statusText = "Couldn't reach Hugh."
        error = msg
    }
}

File: lifebook/android/app/src/main/java/com/beyondpandora/hughmanatee/ui/conversation/ConversationViewModel.kt
package com.beyondpandora.hughmanatee.ui.conversation

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.beyondpandora.hughmanatee.data.livekit.HughLiveKitClient
import com.beyondpandora.hughmanatee.data.local.HughDatabase
import com.beyondpandora.hughmanatee.data.local.entities.SessionEntity
import com.beyondpandora.hughmanatee.data.repository.ConversationRepository
import com.beyondpandora.hughmanatee.data.repository.SessionRepository
import com.beyondpandora.hughmanatee.data.worker.AgentConfigRequest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class ConvStatus {
    Loading, Ready, Live, Ending, Error
}

data class ConversationUiState(
    val status: ConvStatus = ConvStatus.Loading,
    val statusText: String = "Finding Hugh…",
    val firstTurn: String? = null,
    val error: String? = null,
    val turnCount: Int = 0,
)

class ConversationViewModel(application: Application) : AndroidViewModel(application) {

    private val db = HughDatabase.getInstance(application)
    private val profileRepo = com.beyondpandora.hughmanatee.data.repository.ProfileRepository(db)
    private val sessionRepo = SessionRepository(db)
    private val convRepo = ConversationRepository()
    private val liveKit = HughLiveKitClient(application)

    private val _state = MutableStateFlow(ConversationUiState())
    val state: StateFlow<ConversationUiState> = _state.asStateFlow()

    private var currentSessionId: String? = null
    private var turnOrdinal = 0
    private val turnsBuffer = mutableListOf<Pair<String, String>>() // (speaker, text)

    private var started = false
    private var ended = false

    fun startSession() {
        if (started || ended) return
        started = true

        viewModelScope.launch {
            try {
                val profile = profileRepo.get() ?: run {
                    _state.value = _state.value.copy(
                        status = ConvStatus.Error,
                        error = "No profile found",
                    )
                    return@launch
                }

                val lastAnchor = sessionRepo.getLastAnchor()

                val cfg = convRepo.fetchAgentConfig(
                    AgentConfigRequest(
                        first_name = profile.firstName,
                        birth_year = profile.birthYear,
                        hometown = profile.hometown,
                        voice_id = profile.voiceId,
                        last_anchor = lastAnchor,
                    )
                ).getOrThrow()

                _state.value = _state.value.copy(
                    status = ConvStatus.Ready,
                    statusText = "Opening the mic…",
                    firstTurn = cfg.first_turn,
                )

                // Create session row in local DB
                val session = sessionRepo.createSession("agent-2026-05-06")
                currentSessionId = session.id

                // Connect LiveKit
                liveKit.connect(cfg.conversation_token)

                // Collect turns
                launch {
                    liveKit.turns.collect { turn ->
                        turnsBuffer.add(turn.speaker to turn.text)
                        val sid = currentSessionId ?: return@collect
                        turnOrdinal++
                        sessionRepo.appendTurn(
                            sessionId = sid,
                            speaker = turn.speaker,
                            text = turn.text,
                            ordinal = turnOrdinal,
                        )
                        _state.value = _state.value.copy(
                            turnCount = turnOrdinal,
                        )
                    }
                }

                // Watch connection status
                launch {
                    liveKit.status.collect { s ->
                        when (s) {
                            is HughLiveKitClient.Status.Connected -> {
                                _state.value = _state.value.copy(
                                    status = ConvStatus.Live,
                                    statusText = "Hugh is listening.",
                                )
                            }
                            is HughLiveKitClient.Status.Error -> {
                                _state.value = _state.value.copy(
                                    status = ConvStatus.Error,
                                    error = s.message,
                                )
                            }
                            is HughLiveKitClient.Status.Disconnected -> {
                                // Handled by finishSession()
                            }
                            else -> {}
                        }
                    }
                }

                // Watch agent speaking state
                launch {
                    liveKit.agentSpeaking.collect { speaking ->
                        if (_state.value.status == ConvStatus.Live) {
                            _state.value = _state.value.copy(
                                statusText = if (speaking) "Hugh is speaking." else "Hugh is listening.",
                            )
                        }
                    }
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    status = ConvStatus.Error,
                    statusText = "Couldn't reach Hugh.",
                    error = e.message ?: "Unknown error",
                )
            }
        }
    }

    fun endSession() {
        if (ended) return
        ended = true
        _state.value = _state.value.copy(status = ConvStatus.Ending, statusText = "Saving the memory…")

        viewModelScope.launch {
            try {
                liveKit.disconnect()
            } catch (_: Exception) {}

            val sid = currentSessionId ?: return@launch

            var title: String? = null
            var anchor: String? = null

            // Fetch anchor from Worker
            try {
                val tail = turnsBuffer.takeLast(10)
                if (tail.isNotEmpty()) {
                    val anchorRes = convRepo.fetchSessionAnchor(tail).getOrNull()
                    title = anchorRes?.title_suggestion
                    anchor = anchorRes?.anchor_phrase
                }
            } catch (_: Exception) {}

            sessionRepo.endSession(
                sessionId = sid,
                title = title,
                anchorPhrase = anchor,
            )
        }
    }

    fun setMuted(muted: Boolean) {
        liveKit.setMuted(muted)
    }

    override fun onCleared() {
        super.onCleared()
        if (!ended) {
            viewModelScope.launch { liveKit.disconnect() }
        }
    }
}
